### ISSUE TITLE
When creating my own FormatEvent type, I have to duplicate a lot of code

### ISSUE THREAD

**Issue by cecton** [tokio-rs/tracing#1039]
### Crates

tracing-subscriber

### Motivation

I'm trying to customize the log lines in the terminal. For that I need to write my own [`FormatEvent`](https://docs.rs/tracing-subscriber/0.2.13/tracing_subscriber/fmt/trait.FormatEvent.html).

I copy-pasted the code of the one provided by tracing-subsriber called [`Format`](https://docs.rs/tracing-subscriber/0.2.13/src/tracing_subscriber/fmt/format/mod.rs.html#177). But I had to duplicate a lot of other objects before being able to make it useful: FmtLevel, FmtThreadName, FmtCtx, even [`format::time::write()`](https://docs.rs/tracing-subscriber/0.2.13/src/tracing_subscriber/fmt/time/mod.rs.html#252)

[Example here](https://github.com/paritytech/substrate/pull/7328/files#diff-2aff70c0d7f47bf8bc46d493812750120e424ca014752c1ba8b25c03a759a064)

### Proposal

Not sure. Maybe make some common things `pub`?

### Alternatives

Not sure either.

cc @davidbarsky

**Comment by hawkw** [comment]
Yeah, the current `FormatEvent` interface could use some work. I'd like to prioritize improving this as part of the upcoming `tracing-subscriber` 0.3 release.

**Comment by dvdplm** [comment]
@hawkw let me know if you want any help with this. :)

**Comment by hawkw** [comment]
@dvdplm thanks for the offer! Right now, I think we mostly need to answer some API design questions --- what do we need to add, and what do we need to change? It would be very helpful to get answers to the following:
- What behaviors do users implementing `FormatEvent` want pre-built implementations of?
  + What "knobs" or points for additional customization do those pre-built implementations need to provide to be usable?
- What's hard to use or restrictive about the current API? How can we change it to make it easier to use?

**Comment by cecton** [comment]
I'm putting this link here as an example of what I needed to duplicate: https://github.com/paritytech/substrate/blob/2e7292cd84121db8bcd2317c1ad70e348ee52f7a/client/cli/src/logging.rs

But there is going to be a lot more because now I have a use case where I want to make a `Layer` to replace fmt's [`Layer`](https://docs.rs/tracing-subscriber/0.2.15/tracing_subscriber/fmt/struct.Layer.html). This is because we also run the application in the browser so it should use `console.log()` instead of a `Writer`. There is a crate that does that already (tracing-wasm) but it ignores the `FormatEvent`. See the [issue here](https://github.com/storyscript/tracing-wasm/issues/13) for more details.

In the list of issues I have with the current API:
1.  too many things I need are not `pub` so I have to duplicate quite a lot of code to make my own `FormatEvent`
2.  I can't create my own ["format Layer"](https://docs.rs/tracing-subscriber/0.2.15/tracing_subscriber/fmt/struct.Layer.html) because `FmtContext` is needed in input for `FormatEvent` but it has private fields and no constructor.

**Comment by cecton** [comment]
@hawkw maybe it's far fetched but would it be okay if I make a PR where I simply put some stuff public? That would reduce the size and complexity of my code already. It's not that I mind personally but we are a lot of people to maintain it and it's kinda complex. (Code is [here](https://github.com/paritytech/substrate/blob/b17aa9de9e22aa5ec9bd2909ab19bd6cdc8a6982/client/service/src/logging.rs))

**Comment by hawkw** [comment]
I'm a little wary of making too much of the current interface `pub` --- then, it becomes a breaking change to remove it in the future. Rather than making `FmtContext` public and committing to maintaining it forever, I'd prefer to refactor the API so that the `FmtContext` type isn't necessary (https://github.com/tokio-rs/tracing/issues/658).

**Comment by cecton** [comment]
Makes perfect sense... I'm not sure I will be able to help much on this refactoring. I think it's gonna be pretty deep. But don't hesitate to ping me.

In the source code I linked there are a lot of "TODO" marked with "duplicated code" or "inspired from" so you can see the issue I encountered. I think [this file](https://github.com/paritytech/substrate/blob/2e7292cd84121db8bcd2317c1ad70e348ee52f7a/client/cli/src/logging.rs) is the more representative of the "missing pubs".

On the other hand I also had issues substituting the "format Layer" (the layer that actually writes to the Writer). This is in [this file](https://github.com/paritytech/substrate/blob/b17aa9de9e22aa5ec9bd2909ab19bd6cdc8a6982/client/service/src/logging.rs). Probably what `tracing` needs is giving the user the ability to opt in (or not) of the default "format layer".

Lastly there is the problem of accessing the EventFormat object (also in [that file](https://github.com/paritytech/substrate/blob/b17aa9de9e22aa5ec9bd2909ab19bd6cdc8a6982/client/service/src/logging.rs)). Ideally it should be accessible from any `Layer` so one can implement their own "format layer" and use the EventFormat that has been set.

Last lastly: to be honest --but not entirely related to this issue-- we also encountered an issue with the key-value thingy in the span macros. They allow a few types (u64, i64, String and bool) but in my use case I need to send a HashMap of some sort. In slog they had a more flexible key-value thingy that uses serde's `Serialize` so we could send properly (we were using slog before).

**Comment by dvdplm** [comment]
> Last lastly: to be honest --but not entirely related to this issue-- we also encountered an issue with the key-value thingy in the span macros. They allow a few types (u64, i64, String and bool) but in my use case I need to send a HashMap of some sort. In slog they had a more flexible key-value thingy that uses serde's `Serialize` so we could send properly (we were using slog before).

@cecton Would https://github.com/tokio-rs/tracing/pull/886 help here?

**Comment by cecton** [comment]
@dvdplm  No... I saw it already but I don't think it will :disappointed: The thing I sent is missing a key in the JSON as you can [see here](https://github.com/paritytech/substrate/blob/b17aa9de9e22aa5ec9bd2909ab19bd6cdc8a6982/client/telemetry/src/lib.rs#L337). If we could send the whole thing un-serialized (like it was with slog, [see here](https://github.com/paritytech/substrate/blob/2e7292cd84121db8bcd2317c1ad70e348ee52f7a/client/telemetry/src/lib.rs#L319)) I wouldn't need to serialize early like I did, I would serialize it in the `Layer`'s `on_event` where I have the span's id. My current solution right now is hacky but fast: I serialize to JSON once and then I tweak the String to inject the span's id.

**Comment by hawkw** [comment]
> > Last lastly: to be honest --but not entirely related to this issue-- we also encountered an issue with the key-value thingy in the span macros. They allow a few types (u64, i64, String and bool) but in my use case I need to send a HashMap of some sort. In slog they had a more flexible key-value thingy that uses serde's `Serialize` so we could send properly (we were using slog before).
> 
> @cecton Would #886 help here?

I think the issue that would help here is #819. We're definitely interested in adding `serde` support in the new value system for `tracing` 0.2.

**Comment by awused** [comment]
I wanted to make a relatively simple change (chop off `my_crate::` from my own logs and shorten target strings) but the amount of copied code and extra dependencies - like needing to add `nu-ansi-term` (for replicating styling) and `tracing-log` (for `NormalizedEvent`) - is rather prohibitive. It seems like whatever is going to be done to make this better is going to be a breaking change either way, and it's been a few years, so maybe it's worth reconsidering if things like `FmtLevel` and `Writer::dimmed/bold` really need to stay private?

**Comment by thomaseizinger** [comment]
I have a similar motivation. I wanted to explore with moving the printing of `span`s to the end of the log line instead of at the beginning. That would make logs easier to read because the actual message always starts at more or less the same indentation and additional context follows after that.

**Comment by hormelcookies** [comment]
Being unable to write custom formatters without copying a ton of code, forking the crate, or suffering the performance impact from serializing and reserializing JSON every time we write has, singlehandedly, made this one of the most frustrating logging libraries I've ever had the displeasure of using. Please fix this.

**Comment by cecton** [comment]
> Being unable to write custom formatters without copying a ton of code, forking the crate, or suffering the performance impact from serializing and reserializing JSON every time we write has, singlehandedly, made this one of **the most frustrating logging libraries I've ever had the displeasure of using.** **Please fix this.**

I understand the frustration, but I’d encourage a more constructive tone. The maintainers are offering this crate for free, and respectful feedback helps everyone. It’s not about the technical point — which is valid — but how it’s delivered. There’s always a way to express concerns without discouraging those doing the work.

**Comment by lbeschastny** [comment]
I just tried writing my own `FormatEvent` to add some tracing related metadata to our logs.

While writing a custom `format_event` wasn't that bad, it was very frustrating that I had to copy-paste a lot of other utility structs used by `Format<Json, T>` implementation.

Here are some examples:

 * `tracing_subscriber::fmt::writer::WriteAdaptor`
 * `tracing_subscriber::fmt::format::json::SerializableSpan`
 * `tracing_subscriber::fmt::format::json::SerializableContext`

Can you make those public?

That would be very helpful!

**Comment by thomaseizinger** [comment]
> Can you make those public?

Maybe instead of that, they could be moved to a `tracing_subscriber_utils` crate? That way, they are reusable for others without extending the public API of `tracing_subscriber`.

**Comment by lbeschastny** [comment]
> Maybe instead of that, they could be moved to a `tracing_subscriber_utils` crate? That way, they are reusable for others without extending the public API of `tracing_subscriber`.

That would be even better! 

That way it would be possible to introduce breaking changes to the utility crate without affecting the main one.

**Comment by thomaseizinger** [comment]
> That way it would be possible to introduce breaking changes to the utility crate without affecting the main one.

Yep. Just needs someone to step up and create it. We can always transfer it to the tracing folks later if they want.
